apps/zotonic_mod_translation/src/support/translation_translate.erl

```diff
@@ -0,0 +1,154 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(translation_translate).
+-author("Marc Worrell <marc@worrell.nl>").
+
+-include_lib("zotonic_core/include/zotonic.hrl").
+
+-export([ translate/4 ]).
+
+%% @doc Translate one or more strings from one language to another. The strings
+%% can be trans records, a single string or a list of strings. If the source language
+%% is 'en' then the .po files are consulted for a preferred translation.
+%% Both the from and to language must be configured as editable languages.
+
+-spec translate(FromLanguage, ToLanguage, Texts, Context) -> {ok, Translations} | {error, Reason} when
+    FromLanguage :: z_language:language(),
+    ToLanguage :: z_language:language(),
+    Texts :: [ Text ] | Text,
+    Text :: binary() | #trans{},
+    Context :: z:context(),
+    Translations :: [ binary() | undefined ],
+    Reason :: term().
+translate(FromLanguage, ToLanguage, Texts, Context) ->
+    FromCode = to_lang(FromLanguage, Context),
+    ToCode = to_lang(ToLanguage, Context),
+    translate_1(FromCode, ToCode, to_list(Texts), Context).
+
+translate_1({error, _}, _ToCode, _Texts, _Context) ->
+    {error, from_language};
+translate_1(_FromCode, {error, _}, _Texts, _Context) ->
+    {error, to_language};
+translate_1({ok, FromCode}, {ok, ToCode}, Texts, Context) ->
+    PartlyTranslated = lists:map(
+        fun(Text) ->
+            local_trans(FromCode, ToCode, trim(Text), Context)
+        end,
+        Texts),
+    add_service_trans(FromCode, ToCode, PartlyTranslated, Context).
+
+
+add_service_trans(FromCode, ToCode, PartlyTranslated, Context) ->
+    case lists:filtermap(
+        fun
+            ({FromText, undefined}) -> {true, FromText};
+            (_) -> false
+        end,
+        PartlyTranslated)
+    of
+        [] ->
+            Translations = [ T || {_, T} <- PartlyTranslated ],
+            {ok, Translations};
+        ToTranslate ->
+            Notification = #translate{
+                from = FromCode,
+                to = ToCode,
+                texts = ToTranslate
+            },
+            case z_notifier:first(Notification, Context) of
+                {ok, Translations} ->
+                    {ok, merge_translations(PartlyTranslated, Translations, [])};
+                {error, _} = Error ->
+                    Error;
+                undefined ->
+                    {_, Translations} = lists:unzip(PartlyTranslated),
+                    {ok, Translations}
+            end
+    end.
+
+merge_translations([], [], Acc) ->
+    lists:reverse(Acc);
+merge_translations([{_Text, undefined}|Rest], [Tr|Trs], Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]);
+merge_translations([{_Text, Tr}|Rest], Trs, Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]).
```

Comment:
What happens with the case `[{_Text, undefined}|Rest], []`? Would it add `undefined` to `Acc`? Isn't it clearer to clean up the PartiallyTranslated list first, earlier in the code, because its shape is now handled in three different ways :) with the same result if I am correct.

Also, is it necessary to interleave merged translations?
I might have completely missed an implicit ordering dependency 🤔

Reply:
I have the working assumption here that if we request the translation of N strings that we also get N strings back, and that they are in the same order. This merge is merging using that same assumption.

Reply:
Aha, I wasn't sure that when the internal translations cover only part of the translations, the ordering would not change (because of possible gaps in the list of available translations and requesting a subset (the untranslated translations) from the remote service).